### PR TITLE
【Unsolicited PR】I changed the download method to testable.

### PR DESCRIPTION
**Pull request by ytetsuro**
I think Response::download method is not testable.
because use exit and echo.
my solution is following.

1. create Download Response class.
1. to exit when Download Response class returns from Controller method.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] Conforms to style guide

### CODE REVIEW

**Comment by ytetsuro** [comment]
@lonnieezell 

Thank you for reviewing
I am very very very happy.😻

Certainly I understand it well.
However, when I inherited Response class I felt the following discomfort.

I think in the CodeIgniter class that finally receives the DownloadResponse class, there are many conditional branches that depend on the Response class, and if you inherit the Response class, maintenance will be difficult.
DownloadResponse will have redirect and download methods when extends Response.
I think should not have redirect and response methods in DownloadResponse.

Or should we inherit Response thinking that these are not much problems?

File: system/HTTP/DownloadResponse.php [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+	/**
+	 * set download for binary string.
+	 *
+	 * @param string $binary
+	 */
+	public function setBinary(string $binary)
+	{
+		if ($this->file !== null)
+		{
+			throw new BadMethodCallException('When setting filepath can not set binary.');
```
[the remark this answers is not in the document]
Comment (ytetsuro) [reply system/HTTP/DownloadResponse.php]: Thanks.
fixed.

**Comment by lonnieezell** [comment]
> @lonnieezell
> 
> Thank you for reviewing
> I am very very very happy.😻
> 
> Certainly I understand it well.
> However, when I inherited Response class I felt the following discomfort.
> 
> I think in the `CodeIgniter` class that finally receives the `DownloadResponse` class, there are many conditional branches that depend on the `Response` class, and if i inherit the `Response` class, maintenance will be difficult.
> `DownloadResponse` will have `redirect` and `download` methods when extends Response.
> I think should not have redirect and response methods in `DownloadResponse`.
> 
> Or should we inherit `Response` thinking that these are not much problems?

That's a good point. I looked at the flow there, and you're right it would be simpler to extend Message directly for this. In that case - I can't imagine all of those duplicate methods are actually used to force a download? Please take a look and see how much you can get rid of from the original Response class.

**Comment by ytetsuro** [comment]
> Please take a look and see how much you can get rid of from the original Response class.

Certainly, by inheriting the Respone class you can erase many methods from the DownloadResponse class.
But, i analyzed Response class methods.
The analysis results are as follows.

## Reusable methods.

* pretend
* getStatusCode
* setStatusCode
* getReason
* setDate
* setLastModified
* setContentType

## Method that needs to be changed so that it can not be called.

* setJSON
* getJSON
* setXML
* getXML
* setCache
* getBody
* redirect
* setCookie
* hasCookie
* getCookie
* deleteCookie
* sendCookies
* download

## Methods that need to be overwritten.

* __construct
* noCache
* send
* sendHeaders
* sendBody

## Overwrite is unnecessary because it is closed but DownloadResponse does not need unnecessary methods.

* formatBody

I feel that there are many methods that need to be changed so that they can not be called.
Is it better to inherit the Response class really better?

**Comment by lonnieezell** [comment]
@ytetsuro 

> Is that what you are saying does not need to implement RequestInterface?

I assume you mean `ResponseInterface` not the `RequestInterface`?

If so - we still need to implement [ResponseInterface](https://github.com/bcit-ci/CodeIgniter4/blob/develop/system/HTTP/ResponseInterface.php) - however the interface only requires 9 methods.

**Comment by ytetsuro** [comment]
@lonnieezell 

> I assume you mean ResponseInterface not the RequestInterface?

Sorry...
This is typo.
`ResponseInterface` is my intention.

> If so - we still need to implement ResponseInterface - however the interface only requires 9 methods.

Why do you think need to implements the `ResponseInterface`?

**Comment by lonnieezell** [comment]
> Why do you think need to implements the `ResponseInterface`?

Simply because it is a Response and should be cast as such. While we may or may not do any checks on that currently, we might in the future and no sense not following the interface now. Besides, this should still go through the `after` filters (which I just realized I failed to check in your implementation) before being sent, which requires objects implementing ResponseInterface to work.

**Comment by ytetsuro** [comment]
@lonnieezell

> Simply because it is a Response and should be cast as such. While we may or may not do any checks on that currently, we might in the future and no sense not following the interface now. Besides, this should still go through the after filters (which I just realized I failed to check in your implementation) before being sent, which requires objects implementing ResponseInterface to work.

I understand.
I correct it so that it passes through `after` filter in CodeIgniter class.

but, inheriting the Response class will add 13 unnecessary methods.
I think that it would be better to inherit the Message class even though the maintenance problem is considered, but what do you think?

**Comment by lonnieezell** [comment]
> but, add 13 unnecessary methods when extend the Response class.

All I said is that it should implement the ResponseInterface. It's not necessary to extend the Response class itself. Extending Message is fine, as long as you implement the ResponseInterface interface. That interface only has 9 methods specified, I think. A few might be unnecessary, in which case maybe throwing an exception so the developer can correct his code?

**Comment by ytetsuro** [comment]
> I think. A few might be unnecessary, in which case maybe throwing an exception so the developer can correct his code?

Sure.
Modify it to throwing an exception.

**Comment by ytetsuro** [comment]
@lonnieezell 

All have been corrected.
Would you take a moment to review again.

**Comment by ytetsuro** [comment]
Merged very very very thanks.❤️

**Pull request merged**